### DISCUSSION TITLE
Korean language support

### DISCUSSION THREAD

**Discussion by intoxicated** [meilisearch/product#564] [Feedback & Feature Proposal]
I'm thinking to switch from Algolia, but only if Meilisearch can support Korean language. I'd appreciate if you can consider to add :)

**Comment by ManyTheFish** [comment]
Hello @intoxicated, for now, Korean Language has no specialized segmentation or normalization in Meilisearch.
That means that the words will be split by whitespaces and we don't make any modifications to Korean characters.
That's been said, during the Hacktoberfest a contributor added a [specialized segmenter for the Korean Language](https://github.com/meilisearch/charabia/pull/154) so we expect the Korean words to be better split in future versions.
Therefore, if you want to suggest any modifications or share some examples of what you expect in terms of search results, I would be happy to find a solution to enhance it.

**Comment by intoxicated** [comment]
For instance, nori tokenizer is one of standard tokenizer for korean language in ES. https://www.elastic.co/guide/en/elasticsearch/plugins/current/analysis-nori-tokenizer.html whitespace tokenizing is definitely not enough for production use imo.

**Comment by ManyTheFish** [comment]
Hello @intoxicated,
Next week we will release a new version of Meilisearch including a better segmentation of the Korean Language.
You can already try the last beta version [v1.0.0-rc.3](https://github.com/meilisearch/meilisearch/releases) now.

Don't hesitate to give us feedback about this change and point out the remaining weaknesses of Meilisearch Korean support in order to help us enhance it in future versions.

Thanks!

**Comment by intoxicated** [comment]
as version `1.0.0`, I reported on https://github.com/meilisearch/product/discussions/623 that tokenizer behavior is inaccurate how it parses korean sentences. There are mainly two issues,

1. highlighting is inaccurate (containing not exact match but additional characters)
2. exact match doesn't work (i.e. two documents have searchable attribute with content `감독` and search result shows two items when searching with `감` but only one result is hit when searching with `감독`

**Reply by intoxicated** [reply]
@ManyTheFish I'm using https://github.com/meilisearch/meilisearch-kubernetes/releases/tag/meilisearch-0.1.49 to test on remote server, and I believe the https://github.com/meilisearch/charabia/pull/154 is already included in that version.

**Reply by intoxicated** [reply]
for priority, getting right result is way more important than highlighting (can manually find highlight keyword)

**Reply by intoxicated** [reply]
let me know if I can contribute in any ways! I'm willing to contribute :)

**Reply by ManyTheFish** [reply]
Hello @intoxicated,
you previously suggested [nori](https://www.elastic.co/guide/en/elasticsearch/plugins/current/analysis-nori-tokenizer.html) as a segmenter for Korean, did you try it, and do you have the feeling that it's a good segmenter?
In case you think it is a good replacer, I investigated the algorithm behind it and I found that it's a Viterbi algorithm. Good news because Lindera, the segmenter that we use currently, uses the same algorithm, meaning that we "just" have to find the word dictionary used by nori and use it in Lindera, however, I didn't find the source of this dictionary so far. The first step could be to find a good dictionary for Korean Language segmentation. 🤔 

Thank you for your investment in the project!

**Reply by intoxicated** [reply]
https://www.elastic.co/guide/en/elasticsearch/plugins/current/analysis-nori-tokenizer.html this link states that they use https://bitbucket.org/eunjeon/mecab-ko-dic for default dictionary.

**Reply by intoxicated** [reply]
you can find details about nori from [here](https://www.elastic.co/blog/nori-the-official-elasticsearch-plugin-for-korean-language-analysis)

**Reply by ManyTheFish** [reply]
Mmmmh... 🤔
Unfortunately, [Lindera uses mecab-ko-dic-2.1.1-20180720](https://github.com/lindera-morphology/lindera/blob/main/lindera-ko-dic/NOTICE.txt) too, it seems that the dictionary isn't the issue. 🤔 
We could probably[ add a penalty cost on the word length in Lindera](https://docs.rs/lindera-core/0.23.0/lindera_core/viterbi/struct.Penalty.html) (like 4 or 5) to force it to create smaller words that would be easier to find with Meilisearch, however, I don't manage to find a word length to define based on an average or a median.

I have several questions:
1) Do you have in mind a way to retrieve a word length frequency distribution in Korean? something like that: http://www.ravi.io/language-word-lengths
2) Do you have in mind a normalization process we could implement to enhance Korean support?
3) about your phrase search issue, could you share the 2 documents and the query that should match both documents,  please?

**Reply by intoxicated** [reply]
1. I don't think you can take [this](https://hermit1004language.blogspot.com/2015/07/blog-post.html) info as official, but it gives you somewhat general idea how korean word distribution is, (imo length 2~3 can cover alot of word as you can see in the graph). 
2. not that I know of,, (not an expert on NLP)
3. that was one example, today I found more inaccurate result

```
document A { text: "칼데아가 고객에게 감동을 전달을 전달하려면 어떻게 해야할까...?" }
document B { text: "한국 축구 국가 대표팀 감독으로 으로 클린스만 전 미국 감독이 선임되었다." }
document C { text: "감동을 전달을 전달하려면 어떻게 해야할까...?" }
```

When I searched "감동", document A and document C should be matched with like below (single quote for highlighted text)

```
document A { text: "칼데아가 고객에게 '감동'을 전달을 전달하려면 어떻게 해야할까...?" }
document C { text: "'감동'을 전달을 전달하려면 어떻게 해야할까...?" }
```

but meilisearch gave me 
```
document A { text: "칼데아가 고객에게 '감동을 전달'을 전달하려면 어떻게 해야할까...?" }
document B { text: "한국 축구 국가 대표팀 '감독으로 '으로 클린스만 전 미국 감독이 선임되었다." }
document C { text: "'감동을 전달'을 전달하려면 어떻게 해야할까...?" }
```

so to me it seems like the tokenizer doesn't properly tokenize documents. or is there any option that I need to set in order to make it work?

**Reply by intoxicated** [reply]
This is nori tokenizer example for input "개발자 츄르 입니다"

```
GET _analyze
{
  "tokenizer": "nori_tokenizer",
  "text" : [
      "개발자 츄르 입니다"
    ]
}
# 실행 결과 
{
  "tokens" : [
    {
      "token" : "개발",
      "start_offset" : 0,
      "end_offset" : 2,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "자",
      "start_offset" : 2,
      "end_offset" : 3,
      "type" : "word",
      "position" : 1
    },
    {
      "token" : "츄르",
      "start_offset" : 4,
      "end_offset" : 7,
      "type" : "word",
      "position" : 2
    },
    {
      "token" : "이",
      "start_offset" : 8,
      "end_offset" : 11,
      "type" : "word",
      "position" : 3
    },
    {
      "token" : "ᄇ니다",
      "start_offset" : 8,
      "end_offset" : 11,
      "type" : "word",
      "position" : 4
    }
  ]
}
```

**Reply by intoxicated** [reply]
and not sure this is only for dashboard issue but some contents are modified on the site

![Screenshot 2023-03-09 at 9 08 38 PM](https://user-images.githubusercontent.com/848531/224019029-d87fef3d-2c1a-482f-9c46-3aa094f882f7.png)

I searched for "한국" and content has "한국 축구 국가 대표팀 .." but it appears as 

"'한국 축구' 축구 국가 대표팀 ..." where word "축구" is appeared twice with highlight result, but it supposed to be 
"'한국' 축구 국가 대표팀 ..."

**Reply by ManyTheFish** [reply]
Hello @intoxicated,
thank you for your feedback, let's try to make a prototype pimping the behavior of the tokenizer, I'll not have the time to do it this week but maybe next week.

For the highlighting bug, I see that you are using v1.0.0, please use a version higher or equal than v1.0.2, this should fix this specific case.

**Reply by intoxicated** [reply]
got it, i will test with latest version for highlight issue

**Reply by intoxicated** [reply]
ok I just briefly tested with v1.0.2 and it seems much better than previous version I used

**Comment by intoxicated** [comment]
![Screenshot 2023-03-28 at 7 05 58 PM](https://user-images.githubusercontent.com/848531/228202979-9c072950-eae0-41b2-9533-e5bafd128ac8.png)
![Screenshot 2023-03-28 at 7 06 13 PM](https://user-images.githubusercontent.com/848531/228202988-4108749c-9af8-4e5a-9f3a-b123a2969c21.png)
![Screenshot 2023-03-28 at 7 06 19 PM](https://user-images.githubusercontent.com/848531/228202999-b59493f8-7aa7-4d00-933e-efa60bb43c7c.png)

Today I found this issue. All three records in the pictures contain "타자" (meaning typing here) and but when searching "타" or "타자" only shows one result (which record is happened to be prefix with "타자") @ManyTheFish cc
